packages/jsii/test/deprecation-warnings-transformer.test.ts

```diff
@@ -0,0 +1,425 @@
+import { compileJsiiForTest, HelperCompilationResult } from '../lib';
+
+const DEPRECATED = '/** @deprecated Use something else */';
+
+describe('Deprecation warnings', () => {
+  test('deprecated methods', async () => {
+    const result = await compileJsiiForTest(
+      `
+    export class Foo {
+      ${DEPRECATED}
+      public bar(){}
+    }
+    `,
+      undefined /* callback */,
+      { addDeprecationWarnings: true },
+    );
+
+    expect(jsFile(result)).toMatch(
+      'bar() { printJsiiDeprecationWarnings("testpkg.Foo.bar", "Use something else", ""); }',
+    );
+  });
+
+  test('does not inject warnings by default', async () => {
+    const result = await compileJsiiForTest(
+      `
+    ${DEPRECATED}
+    export class Foo {
+      ${DEPRECATED}
+      public bar(){}
+    }
+    `,
+      undefined /* callback */,
+      {}, // addDeprecationWarnings was not set
+    );
+
+    expect(jsFile(result)).toMatch('bar() { }');
+  });
+
+  test.skip('methods that receive parameters of enum types with deprecated values', async () => {
+    const result = await compileJsiiForTest(
+      `
+      export enum SecurityPolicy {
+        ${DEPRECATED}
+        TLS_1_0 = 'TLS_1_0',
+        TLS_1_2 = 'TLS_1_2',
+      }
+      
+    export class Foo {
+      public bar(policy: SecurityPolicy){return policy;}
+    }
+    `,
+      undefined /* callback */,
+      { addDeprecationWarnings: true },
+    );
+
+    jsFile(result);
+    fail('How should we handle this case?');
```

Comment:
The problem here is this: the third argument to `printJsiiDeprecationWarnings` is the value of the deprecated element, as a de-referencing chain (`foo?.bar?.zee`), starting from the parameter, if applicable. This is used by the function to decide whether to show the warning. For example, let's say in the source code compiled by JSII we have:

```ts
interface FooProps {
  readonly fooName: string,
 
  /** @deprecated Blah blah blah */
  readonly bar: string,
}
```
 
And in the user's application code we have:
 
```ts
new Foo(this, 'foo', {
  fooName: 'myFoo',
});
```

Since they are only passing the `fooName` property, which is not deprecated, no warning should be printed. The solution in this PR is to inject a call like this in the constructor:

```ts
printJsiiDeprecationWarnings("my-package.FooProps.bar", "Blah blah blah", props.bar);
```

Because `props.bar` is `undefined`, the function will not print the warning.

So far, so good. Now if we have deprecated enum elements, like:

```ts
enum MyEnum {
  /** @deprecated */
  P1 = 0,
  P2 = 1
}

class Foo {
   bar(e: MyEnum){...}
}
```
and the user's code is:

```ts
foo.bar(MyEnum.P1); // <-- SHOULD print a warning
foo.bar(MyEnum.P2); // <-- SHOULD NOT print a warning
```
What is the best way to address this? The best solution I can come up with is to generate a call for each deprecated enum member like this:

```ts
printJsiiDeprecationWarnings("my-package.MyEnum.P1", message, e === MyEnum.P1 ? e : undefined);
```

This would work, but my concern is the use of `MyEnum`. Depending on how it was imported, we would need to use different namespaces (or none) to reference it. Any ideas?

Reply:
I think for every type we need a function, and call them recursively:

```ts
enum MyEnum {
  /** @deprecated */
  P1 = 0,
  P2 = 1
}

interface FooProps {
  readonly okay: string;

  /** @deprecated */
  readonly ohno: string;

  readonly enum: MyEnum;
}

class Foo {
   bar(props: FooProps) {
  }
}
```

To my mind should synth to something like this:

```ts
class Foo {
  bar(props) { 
    $_checkDeprecations_FooProps(props, `Foo#bar: fooprops`);
  }
}

function $_checkDeprecations_FooProps(props, path) {
  if (props?.ohno) { warnDeprecation(`${path}.ohno`); }
  $_checkDeprecations_MyEnum(props?.enum, `${path}.enum`);
}

function $_checkDeprecations_MyEnum(enum, path) {
  if (enum === MyEnum.P1) { warnDeprecation(`${path}.P1`); }
}
```

Or something. Functions will save on generated code if the same interface type is referenced in multiple places, and is also the only way to deal with recursive types:

```ts
interface MyStruct {
  thing: string;
  substructs: MyStruct[];
}
```